File: LeetCode - Top Interview 150/199.py
# 199. Binary Tree Right Side View
# https://leetcode.com/problems/binary-tree-right-side-view/


from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return None

        result = []

        queue = []
        queue.append((root, 1))

        while queue:
            node, depth = queue.pop(0)

            if len(result) < depth:
                result.append(node.val)
                print(result)

            if node.right:
                queue.append((node.right, depth + 1))
            if node.left:
                queue.append((node.left, depth + 1))

        return result

File: LeetCode - Algorithm/Binary Tree/199.js
// 199. Binary Tree Right Side View
// https://leetcode.com/problems/binary-tree-right-side-view/

function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var rightSideView = function (root) {
  if (!root || root.length === 0) return [];

  const result = [];

  const queue = [];
  queue.push(root);

  while (queue.length > 0) {
    const currentLength = queue.length;

    let node;

    for (let i = 0; i < currentLength; i++) {
      node = queue.shift();

      if (node.left) queue.push(node.left);
      if (node.right) queue.push(node.right);
    }

    result.push(node.val);
  }

  return result;
};
